sky/backends/cloud_vm_ray_backend.py

```diff
@@ -4130,6 +4132,35 @@ def post_teardown_cleanup(self,
         config = common_utils.read_yaml(handle.cluster_yaml)
         backend_utils.SSHConfigHelper.remove_cluster(handle.cluster_name)
 
+        # Confirm that instances have actually transitioned state before
+        # updating the state database.
+        attempts = 0
+        while True:
+            try:
+                logger.debug(f'instance statuses attempt {attempts + 1}')
+                node_status_dict = provision_lib.query_instances(
+                    repr(cloud),
+                    cluster_name_on_cloud,
+                    config['provider'],
+                    non_terminated_only=False)
+                # FIXME(cooperc): Some clouds (e.g. GCP) do not distinguish
+                # between "stopping/stopped" and "terminating/terminated", so we
+                # allow for either status instead of casing on `terminate`.
+                for node_id, node_status in node_status_dict.items():
+                    logger.debug(f'{node_id} status: {node_status}')
+                    if node_status not in [
+                            None, status_lib.ClusterStatus.STOPPED
+                    ]:
+                        raise RuntimeError(f'Instance {node_id} in unexpected '
+                                           f'state {node_status}.')
+                break
+            except RuntimeError:
+                attempts += 1
+                if attempts < _TEARDOWN_WAIT_MAX_ATTEMPTS:
+                    time.sleep(_TEARDOWN_WAIT_BETWEEN_ATTEMPS_SECONDS)
+                else:
+                    raise
```

Comment:
Hmm, I was thinking we have this kind of wait and retry within the `instance.py::terminate_instance/stop_instance`. Any reason we don't have it in the cloud implementation?

Reply:
The idea was to put this as close as possible to the actual state database update, since the termination and the state cleanup are currently somewhat separated.

Reply:
e.g. if this code
https://github.com/skypilot-org/skypilot/blob/6e5083293f0d9a9d069d51274c57f0e59e47e5ce/sky/provision/provisioner.py#L148-L159
throws an exception, then we will not call teardown_cluster, but we will call post_teardown_cleanup here:
https://github.com/skypilot-org/skypilot/blob/6e5083293f0d9a9d069d51274c57f0e59e47e5ce/sky/backends/cloud_vm_ray_backend.py#L1574-L1578

Obviously we don't expect this to happen but keeping the check right in post_teardown_cleanup will ensure that it fails in that case.